docker/pypi/wmagent/init.sh

```diff
@@ -299,15 +301,20 @@ set_cronjob() {
     stepMsg="Populating cronjob with utilitarian scripts for the $WMA_USER"
     echo "-----------------------------------------------------------------------"
     echo "Start: $stepMsg"
-
+    local errVal=0
     chmod +x $WMA_DEPLOY_DIR/deploy/renew_proxy.sh $WMA_DEPLOY_DIR/deploy/restartComponent.sh
 
     crontab -u $WMA_USER - <<EOF
+$(crontab -l)
```

Comment:
With this line, won't we duplicate some of the cron entries upon multiple executions of this init.sh script?

Reply:
This is true, but it is also true that on every container restart we lose all previous cronjobs, so at the time when `init.sh` is executed for the first time during container restart the list of cronjobs is already empty. The `init.sh` script is mostly to be executed when the container starts, but indeed there is the possibility for someone to do it manually. Me personally, many times, if the initialization process fails for some known reason, I just login into the container fix the problem and instead of restarting the container I execute: 
```
$WMA_ROOT_DIR/init.sh
``` 
and let it continue from where it has stopped last time. Said that, we need to address also the fact that those CouchDB related cronjobs are created  only once during the very first steps of the initialization, here: https://github.com/dmwm/CMSKubernetes/blob/c2b643cb95e502c7773e16ab6953c9b920476e60/docker/pypi/wmagent/bin/manage#L153 and the call during the execution of `init.sh` is here: https://github.com/dmwm/CMSKubernetes/blob/c2b643cb95e502c7773e16ab6953c9b920476e60/docker/pypi/wmagent/init.sh#L555    
While we  populate the proxy related cronjobs every time we restart  the container: https://github.com/dmwm/CMSKubernetes/blob/c2b643cb95e502c7773e16ab6953c9b920476e60/docker/pypi/wmagent/init.sh#L579 , we tend to lose the couchdb related ones on restarts.   

So one way would be to add also the `wmagent-couchapp-init` to the call of `set_cronjob`, not a problem at all, but this would push the couch apps yet again. Won't hirt of course, but not needed. What I'd suggest is also to ad an option ot `wmagent-couchapp-init` to only populate the cronjobs`
This would require a change in: https://github.com/dmwm/WMCore/blob/master/bin/wmagent-couchapp-init

Reply:
I actually have a different suggestion then. How about moving the `set_cronjob` method inside the `check_docker_init` block, here:
https://github.com/dmwm/CMSKubernetes/blob/b4fad26ebf59010a6dfe628e8f0600c243758f79/docker/pypi/wmagent/init.sh#L558

such that we run it only once, similar to the CouchDB cron entries. I still don't think it will offer 100% protection against someone running init.sh multiple times, but it will be better for mainly 2 reasons:
1) there is no need to redefine the cron entries upon each init.sh execution
2) it gets to run before the script says that everything was Okay, in this line: https://github.com/dmwm/CMSKubernetes/blob/b4fad26ebf59010a6dfe628e8f0600c243758f79/docker/pypi/wmagent/init.sh#L583

Reply:
All this is really good. I agree with you ... the problem is the cronjobs are not preserved upon container restart - and this is yet another problem we need to fight.
Any time someone gets to restart ac container  he/she is about to end up with an empty cron list, like this : 
```
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ crontab -l 

55 */12 * * * /data/srv/wmagent/2.3.4rc3/config/manage renew-proxy
58 */12 * * * python /usr/local/deploy/checkProxy.py --proxy /data/certs/myproxy.pem --time 120 --send-mail True --mail alan.malta@cern.ch
*/15 * * * *  source /usr/local/deploy/restartComponent.sh > /dev/null
* * * * * echo 'http://***:***@127.0.0.1:5984/wmagent_jobdump\%2Fjobs/_design/JobDump/_view/statusByWorkflowName?limit=1' | sed -e 's|\\||g' | xargs curl -s -m 1 > /dev/null
* * * * * echo 'http://**:**@127.0.0.1:5984/wmagent_jobdump\%2Ffwjrs/_design/FWJRDump/_view/outputByJobID?limit=1' | sed -e 's|\\||g' | xargs curl -s -m 1 > /dev/null
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
exit


cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ docker  kill wmagent 
wmagent
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ ./wmagent-docker-run.sh -t 2.3.4rc3 && docker logs -f wmagent 
Checking if there is no other wmagent container running and creating a link to the 2.3.4rc3 in the host mount area.
Starting wmagent:2.3.4rc3 docker container with user: cmst1:zh
937f912fa102744f9388905b5f7f89c93773d7fea48c7916cf229349a444292a
Running WMAgent container with user: cmst1 (ID: 31961) and group: zh (ID: 1399)
Setting up bashrc for user: cmst1 under home directory: /home/cmst1
Start initialization

=======================================================
Starting WMAgent with the following initialisation data:
-------------------------------------------------------
 - WMAgent Version            : 2.3.4rc3
 - WMAgent User               : cmst1
 - WMAgent Root path          : /data
 - WMAgent Host               : vocms0260.cern.ch
 - WMAgent TeamName           : testbed-vocms0260
 - WMAgent Number             : 0
 - WMAgent Relational DB type : mysql
 - Python  Version            : Python 3.8.16
 - Python  Module path        : /usr/local/lib/python3.8/site-packages
=======================================================

-------------------------------------------------------
Start: Performing basic_checks

Done: Performing basic_checks
-------------------------------------------------------

check_wmasecrets: Checking for changes in the WMAgent.secrets file
check_wmasecrets: No change found.
-------------------------------------------------------
Start: Performing checks for successful Docker initialisation steps...
WMA_BUILD_ID: 98ab24af6b14feb6e15efc6e185f0c0a7275fa87a9c6d29b3248376f2a778c3a
dockerInitId: 98ab24af6b14feb6e15efc6e185f0c0a7275fa87a9c6d29b3248376f2a778c3a
OK
-----------------------------------------------------------------------

_check_mysql: Checking whether the MySQL server is reachable...
_status_of_mysql:
Uptime: 92257  Threads: 17  Questions: 6991092  Slow queries: 0  Opens: 511  Open tables: 84  Queries per second avg: 75.778
_status_of_mysql: MySQL connection is OK!
_check_mysql: Checking whether the MySQL schema has been installed
_sql_schema_valid: Checking the current SQL Database schema integrity.
_sql_dumpSchema: Dumping the current SQL schema of database: wmagent to /tmp/.wmaSchemaTmp
_sql_dbid_valid: Checking if the current SQL Database Id matches the WMA_BUILD_ID and hostname of the agent.
_sql_dbid_valid: OK: Database recorded and current agent's init parameters match.
_check_couch: Checking whether the CouchDB database is reachable...
_status_of_couch:
{"couchdb":"Welcome","version":"3.2.2","git_sha":"d5b746b7c","uuid":"18f53118737ed74893055db0ffa972e2","features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"]}
_status_of_couch: CouchDB connection is OK!
MyProxy v6.2 Aug 2019 PAM SASL KRB5 LDAP VOMS OCSP
Attempting to connect to 2001:1458:d00:4a::100:480:7512 
Successfully connected to myproxy.cern.ch:7512 
using trusted certificates directory /etc/grid-security/certificates
Using Host cert file (/data/certs/servicecert.pem), key file (/data/certs/servicekey.pem)
server name: /DC=ch/DC=cern/OU=computers/CN=px501.cern.ch
checking that server name is acceptable...
server name matches "myproxy.cern.ch"
authenticated server name is acceptable
A credential has been received for user amaltaro in /data/certs/mynewproxy.pem.
Your identity: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=amaltaro/CN=718748/CN=Alan Malta Rodrigues/CN=928368380/CN=2079557601/CN=930553887
Contacting  lcg-voms2.cern.ch:15002 [/DC=ch/DC=cern/OU=computers/CN=lcg-voms2.cern.ch] "cms" Done
Creating proxy  Done

Your proxy is valid until Thu May 23 11:49:41 2024
-------------------------------------------------------
Start: Performing start_agent
-------------------------------------------------------
...
-------------------------------------------------------
Start sleeping now ...zzz...

cmst1@vocms0260:wmagent $ docker exec -it wmagent bash 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ crontab -l 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 

```

Reply:
I have a suggestion how to resolve it. My next commit is about to be pushed in few minutes

Reply:
Ok, with this commit: https://github.com/dmwm/CMSKubernetes/pull/1476/commits/95586e5fea4517e6ac318027e003c5ae239e674e

The problem with the lost cronjobs is fixed. Here is the output of a  cold restart of a freshly initialized agent [1]. In addition I think it is better to have the $WMA_CONFIG_FILE variable exported from the top. It is quite helpful in day operations. It have been many times, when I have to type `vi $WMA_CONFIG_DIR/config.py` to think about this. And it also fits with the rest of the wmagent related environment variable (unwritten) convention to have the `WMA_*` prefix.


[1]
```
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ crontab -l 
55 */12 * * * /data/srv/wmagent/2.3.4rc3/config/manage renew-proxy
58 */12 * * * python /usr/local/deploy/checkProxy.py --proxy /data/certs/myproxy.pem --time 120 --send-mail True --mail alan.malta@cern.ch
*/15 * * * *  source /usr/local/deploy/restartComponent.sh > /dev/null
* * * * * echo 'http://***:***@127.0.0.1:5984/wmagent_jobdump\%2Fjobs/_design/JobDump/_view/statusByWorkflowName?limit=1' | sed -e 's|\\||g' | xargs curl -s -m 1 > /dev/null
* * * * * echo 'http://***:***@127.0.0.1:5984/wmagent_jobdump\%2Ffwjrs/_design/FWJRDump/_view/outputByJobID?limit=1' | sed -e 's|\\||g' | xargs curl -s -m 1 > /dev/null
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ 
exit
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ docker  kill wmagent 
wmagent
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ ./wmagent-docker-run.sh -t 2.3.4rc3 && docker logs -f wmagent 
Checking if there is no other wmagent container running and creating a link to the 2.3.4rc3 in the host mount area.
Starting wmagent:2.3.4rc3 docker container with user: cmst1:zh
144b7659f8e0e590a249d90575c2ce9c689d715744af396ed552903c528e4292
Running WMAgent container with user: cmst1 (ID: 31961) and group: zh (ID: 1399)
Setting up bashrc for user: cmst1 under home directory: /home/cmst1
Start initialization

=======================================================
Starting WMAgent with the following initialisation data:
-------------------------------------------------------
 - WMAgent Version            : 2.3.4rc3
 - WMAgent User               : cmst1
 - WMAgent Root path          : /data
 - WMAgent Host               : vocms0260.cern.ch
 - WMAgent TeamName           : testbed-vocms0260
 - WMAgent Number             : 0
 - WMAgent Relational DB type : mysql
 - Python  Version            : Python 3.8.16
 - Python  Module path        : /usr/local/lib/python3.8/site-packages
=======================================================

-------------------------------------------------------
Start: Performing basic_checks

Done: Performing basic_checks
-------------------------------------------------------

check_wmasecrets: Checking for changes in the WMAgent.secrets file
check_wmasecrets: No change found.
-------------------------------------------------------
Start: Performing checks for successful Docker initialisation steps...
WMA_BUILD_ID: 8ce57b7bfa38f674e67211baaa8361aad212b08d7317c4b0152fc0ea9939e703
dockerInitId: 8ce57b7bfa38f674e67211baaa8361aad212b08d7317c4b0152fc0ea9939e703
OK
-----------------------------------------------------------------------
Start: Populating cronjob with utilitarian scripts for user: cmst1
Installing FWJRDump into wmagent_jobdump/fwjrs
Installing FWJRDump app into database: http://127.0.0.1:5984/wmagent_jobdump%2Ffwjrs
Installing JobDump into wmagent_jobdump/jobs
Installing JobDump app into database: http://127.0.0.1:5984/wmagent_jobdump%2Fjobs
Installing WMStatsAgent into wmagent_summary
Installing WMStatsAgent app into database: http://127.0.0.1:5984/wmagent_summary
Installing SummaryStats into stat_summary
Installing SummaryStats app into database: http://127.0.0.1:5984/stat_summary
Setting up cron jobs for the job dump.
Installing WorkQueue into workqueue
Installing WorkQueue app into database: http://127.0.0.1:5984/workqueue
Installing WorkQueue into workqueue_inbox
Installing WorkQueue app into database: http://127.0.0.1:5984/workqueue_inbox
Done: Populating cronjob with utilitarian scripts for user: cmst1!

-----------------------------------------------------------------------
_check_mysql: Checking whether the MySQL server is reachable...
_status_of_mysql:
Uptime: 95298  Threads: 17  Questions: 7215499  Slow queries: 0  Opens: 976  Open tables: 82  Queries per second avg: 75.715
_status_of_mysql: MySQL connection is OK!
_check_mysql: Checking whether the MySQL schema has been installed
_sql_schema_valid: Checking the current SQL Database schema integrity.
_sql_dumpSchema: Dumping the current SQL schema of database: wmagent to /tmp/.wmaSchemaTmp
_sql_dbid_valid: Checking if the current SQL Database Id matches the WMA_BUILD_ID and hostname of the agent.
_sql_dbid_valid: OK: Database recorded and current agent's init parameters match.
_check_couch: Checking whether the CouchDB database is reachable...
_status_of_couch:
{"couchdb":"Welcome","version":"3.2.2","git_sha":"d5b746b7c","uuid":"18f53118737ed74893055db0ffa972e2","features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"]}
_status_of_couch: CouchDB connection is OK!
MyProxy v6.2 Aug 2019 PAM SASL KRB5 LDAP VOMS OCSP
Attempting to connect to 2001:1458:d00:4a::100:480:7512 
Successfully connected to myproxy.cern.ch:7512 
using trusted certificates directory /etc/grid-security/certificates
Using Host cert file (/data/certs/servicecert.pem), key file (/data/certs/servicekey.pem)
server name: /DC=ch/DC=cern/OU=computers/CN=px501.cern.ch
checking that server name is acceptable...
server name matches "myproxy.cern.ch"
authenticated server name is acceptable
A credential has been received for user amaltaro in /data/certs/mynewproxy.pem.
Your identity: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=amaltaro/CN=718748/CN=Alan Malta Rodrigues/CN=928368380/CN=2079557601/CN=882239323
Contacting  lcg-voms2.cern.ch:15002 [/DC=ch/DC=cern/OU=computers/CN=lcg-voms2.cern.ch] "cms" Done
Creating proxy  Done

Your proxy is valid until Thu May 23 12:40:22 2024
-------------------------------------------------------
Start: Performing start_agent
-------------------------------------------------------
Shutting down WMAgent...
Checking default database connection... ok.
Stopping components: ['WorkQueueManager', 'DBS3Upload', 'JobAccountant', 'JobCreator', 'JobSubmitter', 'JobTracker', 'JobStatusLite', 'JobUpdater', 'ErrorHandler', 'RetryManager', 'JobArchiver', 'TaskArchiver', 'AnalyticsDataCollector', 'ArchiveDataReporter', 'AgentStatusWatcher', 'RucioInjector', 'WorkflowUpdater']
Stopping: WorkQueueManager
Component WorkQueueManager with process id 2398 is not running
Stopping: DBS3Upload
Component DBS3Upload with process id 2407 is not running
Stopping: JobAccountant
Component JobAccountant with process id 2424 is not running
Stopping: JobCreator
Component JobCreator with process id 2432 is not running
Stopping: JobSubmitter
Component JobSubmitter with process id 2440 is not running
Stopping: JobTracker
Component JobTracker with process id 2454 is not running
Stopping: JobStatusLite
Component JobStatusLite with process id 2463 is not running
Stopping: JobUpdater
Component JobUpdater with process id 2472 is not running
Stopping: ErrorHandler
Component ErrorHandler with process id 2481 is not running
Stopping: RetryManager
Component RetryManager with process id 2490 is not running
Stopping: JobArchiver
Component JobArchiver with process id 2498 is not running
Stopping: TaskArchiver
Component TaskArchiver with process id 2507 is not running
Stopping: AnalyticsDataCollector
Component AnalyticsDataCollector with process id 2523 is not running
Stopping: ArchiveDataReporter
Component ArchiveDataReporter with process id 2533 is not running
Stopping: AgentStatusWatcher
Component AgentStatusWatcher with process id 2541 is not running
Stopping: RucioInjector
Component RucioInjector with process id 2630 is not running
Stopping: WorkflowUpdater
Component WorkflowUpdater with process id 2639 is not running
Starting WMAgent...
...........................................................................................
started with pid 785

started with pid 794

started with pid 811

started with pid 819

started with pid 827

started with pid 841

started with pid 850

started with pid 859

started with pid 868

started with pid 877

started with pid 885

started with pid 894

started with pid 910

started with pid 920

started with pid 928

started with pid 962

started with pid 1030
Checking default database connection... ok.
Starting components: ['WorkQueueManager', 'DBS3Upload', 'JobAccountant', 'JobCreator', 'JobSubmitter', 'JobTracker', 'JobStatusLite', 'JobUpdater', 'ErrorHandler', 'RetryManager', 'JobArchiver', 'TaskArchiver', 'AnalyticsDataCollector', 'ArchiveDataReporter', 'AgentStatusWatcher', 'RucioInjector', 'WorkflowUpdater']
Starting : WorkQueueManager
Starting WorkQueueManager as a daemon 
Log will be in /data/srv/wmagent/current/install/WorkQueueManager 
Waiting 1 seconds, to ensure daemon file is created
Starting : DBS3Upload
DBS3Upload.__init__
Starting DBS3Upload as a daemon 
Log will be in /data/srv/wmagent/current/install/DBS3Upload 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobAccountant
Starting JobAccountant as a daemon 
Log will be in /data/srv/wmagent/current/install/JobAccountant 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobCreator
JobCreator.__init__
Starting JobCreator as a daemon 
Log will be in /data/srv/wmagent/current/install/JobCreator 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobSubmitter
Starting JobSubmitter as a daemon 
Log will be in /data/srv/wmagent/current/install/JobSubmitter 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobTracker
JobTracker.__init__
Starting JobTracker as a daemon 
Log will be in /data/srv/wmagent/current/install/JobTracker 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobStatusLite
JobStatusLite.__init__
Starting JobStatusLite as a daemon 
Log will be in /data/srv/wmagent/current/install/JobStatusLite 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobUpdater
Starting JobUpdater as a daemon 
Log will be in /data/srv/wmagent/current/install/JobUpdater 
Waiting 1 seconds, to ensure daemon file is created
Starting : ErrorHandler
Starting ErrorHandler as a daemon 
Log will be in /data/srv/wmagent/current/install/ErrorHandler 
Waiting 1 seconds, to ensure daemon file is created
Starting : RetryManager
Starting RetryManager as a daemon 
Log will be in /data/srv/wmagent/current/install/RetryManager 
Waiting 1 seconds, to ensure daemon file is created
Starting : JobArchiver
JobArchiver.__init__
Starting JobArchiver as a daemon 
Log will be in /data/srv/wmagent/current/install/JobArchiver 
Waiting 1 seconds, to ensure daemon file is created
Starting : TaskArchiver
TaskArchiver.__init__
Starting TaskArchiver as a daemon 
Log will be in /data/srv/wmagent/current/install/TaskArchiver 
Waiting 1 seconds, to ensure daemon file is created
Starting : AnalyticsDataCollector
Starting AnalyticsDataCollector as a daemon 
Log will be in /data/srv/wmagent/current/install/AnalyticsDataCollector 
Waiting 1 seconds, to ensure daemon file is created
Starting : ArchiveDataReporter
Starting ArchiveDataReporter as a daemon 
Log will be in /data/srv/wmagent/current/install/ArchiveDataReporter 
Waiting 1 seconds, to ensure daemon file is created
Starting : AgentStatusWatcher
Starting AgentStatusWatcher as a daemon 
Log will be in /data/srv/wmagent/current/install/AgentStatusWatcher 
Waiting 1 seconds, to ensure daemon file is created
Starting : RucioInjector
Starting RucioInjector as a daemon 
Log will be in /data/srv/wmagent/current/install/RucioInjector 
Waiting 1 seconds, to ensure daemon file is created
Starting : WorkflowUpdater
Starting WorkflowUpdater as a daemon 
Log will be in /data/srv/wmagent/current/install/WorkflowUpdater 
Waiting 1 seconds, to ensure daemon file is created
Done: Performing start_agent
-------------------------------------------------------
Start sleeping now ...zzz...

^C
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ 
cmst1@vocms0260:wmagent $ docker exec -it wmagent bash 
(WMAgent-2.3.4rc3) [cmst1@vocms0260:current]$ crontab -l 
55 */12 * * * /data/srv/wmagent/2.3.4rc3/config/manage renew-proxy
58 */12 * * * python /usr/local/deploy/checkProxy.py --proxy /data/certs/myproxy.pem --time 120 --send-mail True --mail alan.malta@cern.ch
*/15 * * * *  source /usr/local/deploy/restartComponent.sh > /dev/null
* * * * * echo 'http://***:***@127.0.0.1:5984/wmagent_jobdump\%2Fjobs/_design/JobDump/_view/statusByWorkflowName?limit=1' | sed -e 's|\\||g' | xargs curl -s -m 1 > /dev/null
* * * * * echo 'http://***:***@127.0.0.1:5984/wmagent_jobdump\%2Ffwjrs/_design/FWJRDump/_view/outputByJobID?limit=1' | sed -e 's|\\||g' | xargs curl -s -m 1 > /dev/null

```